README.md

```diff
@@ -1,45 +1,263 @@
+
+    # in any process: whenever a document is dirtied
+    Ick.ickadd(redis,queue_key,Time.now.to_f,document_id)
+
+    # in the indexer process:
+    batch = Ick.ickreserve(redis,queue_key,batch_size)
+    process_batch_slowly(batch)
+    # burn down the queue only if the batch succeeded
+    Ick.ickcommit(redis,queue_key,*members_of(batch))   # critical section gone
+
+
+Ick solves for failover via a two phase commit protocol between
+**ickreserve** and **ickcommit**.  If there is a failure during
+process_batch_slowly(batch), the next time time we call **ickreserve**
+we will just get the same batch - it will have resided unchanged in
+the consumer set until we get happy and call **ickcommit**.
+
+Ick solves the Forgotten Dirtiness Problem by virtue of
+**ickreserve**’s implicit atomicity and the fact that **ickcommit** is
+only ever called from the indexer and producers do not mutate the
+consumer set.
+
+Ick solves the Hot Data Starvation Problem by a subtle change in
+ICKADD.  Unlike ZADD, which overwrites the old score when a message is
+re-added, or ZADD NX which always preserves the old score, ICKADD
+always takes the _min_ of the old and new scores.  Thus, Ick tracks
+the first-known ditry time for a message even when there is time skew
+in the producers.  The longer entries stay in the consumer set, the
+more they implicitly percolate toward the cold end regardless of how
+many updates they receive.  Ditto in the consumer set.  Provided that
+all producers make a best effort to use only current or future
+timestamps when they call ICKADD, the ICKRESERVE batch will always
+include the oldest entries and there will be no starvation.
+
+Apology: I know that [Two-Phase
+Commit](https://en.wikipedia.org/wiki/Two-phase_commit_protocol) has a
+different technical meaning than what Ick does.  Unfortunately I can't
+find a better name for this very common failsafe queue pattern.  I
+suppose we could the Redis sorted set as the coordinator and the
+consumer process as the (single) participant node and, generously,
+Two-Phase Commit might be taken to describe Ick.
+
+
+## What is Ick?
+
+An Ick is a collection of three Redis keys which all live on the same
+[Redis hash slot](https://redis.io/topics/cluster-spec):
+
+* version flag, a string
+* producer set, a sorted set into which we flag keys as dirty with timestamps
+* consumer set, a sorted set from which the indexer pulls batches to index
+
+### Ick defines 5 operations on this data via Lua on Redis:
+
+* **ickdel**: removes all keys associated with a given Ick structure
+* **ickstats**: returns a hash of stats including version and size
+* **ickadd**: add a batch of members with scores to the producer set
+** implements write-folding: a message can only appear once in the producer set
+** when a member is re-added, it takes the lowest of 2 scores
+* **ickreserve**: moves members from to the producer set to the consumer set until the consumer set is size N or the producer set is empty
```

Comment:
Looks like this line breaks your established column width.

Reply:
I had thought that Markdown would lose the plot on the bullet point if I included a newline.

I'll do a test and see if I can wrap this line so it looks good in both ASCII and in the processed Markdown.